### PR TITLE
feat(core): add LweCiphertextDiscardingPublicKeyEncryptionEngine

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves:

closes https://github.com/zama-ai/concrete-core-internal/issues/271

### Description

- add spec and impls

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [x] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    /// use concrete_commons::dispersion::Variance;
+    /// use concrete_commons::parameters::{LweDimension, LwePublicKeyZeroEncryptionCount};
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // DISCLAIMER: the parameters used here are only for test purpose, and are not secure.
+    /// let lwe_dimension = LweDimension(2);
+    /// let lwe_public_key_zero_encryption_count = LwePublicKeyZeroEncryptionCount(7);
+    /// // Here a hard-set encoding is applied (shift by 20 bits)
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs:113]: Here we could shift by 50 bits like in other doc tests for u64

**Review by agnesLeroy — APPROVED** [review]
Looks good to me. Shouldn't we implement the "pure" version of the engine as well, and not just the discarding one?

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs:113]: yes, copy paste

**Comment by IceTDrinker** [comment]
can do !

**Comment by IceTDrinker** [comment]
Right !

worst case we may generate the others ;)

File: concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs:113]: done !

Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs:113]: 20 -> 50 in the comment :sob:

Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_engine/lwe_ciphertext_discarding_public_key_encryption.rs:113]: arg ^^ thanks

**Pull request merged**